api/apikeys.go

```diff
@@ -3,16 +3,107 @@ package api
 import (
 	"net/http"
 	"strconv"
+	"time"
 
 	"github.com/SkynetLabs/skynet-accounts/database"
 	"github.com/julienschmidt/httprouter"
 	"gitlab.com/NebulousLabs/errors"
+	"go.mongodb.org/mongo-driver/bson/primitive"
 	"go.mongodb.org/mongo-driver/mongo"
 )
 
+type (
+	// APIKeyPOST describes the body of a POST request that creates an API key
+	APIKeyPOST struct {
+		Public   bool     `json:"public,string"`
+		Skylinks []string `json:"skylinks"`
+	}
+	// APIKeyPUT describes the request body for updating an API key
+	APIKeyPUT struct {
+		Skylinks []string
+	}
+	// APIKeyPATCH describes the request body for updating an API key by
+	// providing only the requested changes
+	APIKeyPATCH struct {
+		Add    []string
+		Remove []string
+	}
+	// APIKeyResponse is an API DTO which mirrors database.APIKey.
+	APIKeyResponse struct {
+		ID        primitive.ObjectID `json:"id"`
+		UserID    primitive.ObjectID `json:"-"`
+		Public    bool               `json:"public,string"`
+		Key       database.APIKey    `json:"-"`
+		Skylinks  []string           `json:"skylinks"`
+		CreatedAt time.Time          `json:"createdAt"`
+	}
+	// APIKeyResponseWithKey is an API DTO which mirrors database.APIKey but
+	// also reveals the value of the Key field. This should only be used on key
+	// creation.
+	APIKeyResponseWithKey struct {
+		APIKeyResponse
+		Key database.APIKey `json:"key"`
+	}
+)
+
+// Validate checks if the request and its parts are valid.
+func (akp APIKeyPOST) Validate() error {
+	if !akp.Public && len(akp.Skylinks) > 0 {
+		return errors.New("public API keys cannot refer to skylinlks")
+	}
+	errs := make([]error, 0)
+	for _, s := range akp.Skylinks {
+		if !database.ValidSkylinkHash(s) {
+			errs = append(errs, errors.New("invalid skylink:"+s))
```

Comment:
```suggestion
			errs = append(errs, errors.New("invalid skylink: "+s))
```
not sure whether the concat operator adds a space by default there?

Reply:
No, it doesn't. `fmt.Println(s1, s2)` adds a space.